### ISSUE TITLE
Aggregating maps for multiple tasks 

### ISSUE THREAD

**Issue by sameerr25** [logstash-plugins/logstash-filter-aggregate#36]
Hey guys,

So, following is part of my logstash conf file:
```
    filter {
        if ( [resource_id] and [resource_type] ) {
            aggregate {
                task_id => "%{resource_type}_%{resource_id}"
                code => "
                    map['tags'] ||= ['aggregated']
                    map['event_count'] ||= 0 ; map['event_count'] += 1 ;
                    event_hash = event.to_hash
                    event.to_hash.each do |key,value|
                        map[key] = value unless map.has_key?(key)
                    end
                "
                push_previous_map_as_event => true
                timeout => 10
            }

            if "aggregated" not in [tags] {
                drop {}
            }
        }
    }
```

When I pass the following 4 values consecutively:
```
    {"resource_id": 1, "resource_type": "flight"}
    {"resource_id": 2, "resource_type": "flight"}
    {"resource_id": 1, "resource_type": "flight"}
    {"resource_id": 2, "resource_type": "flight"}
```

Issue: The value for resource_id 1 gets pushed as event instantly and only the last task with resource_id 2 halts for 10 seconds(timeout) and then gets pushed out, moreover the counter value for each task is not correct both tasks should have counter 2 but instead it's 1, it's like aggregation is not happening.

Expectation: Each task gets it's own hash in code and gets pushed after 10 seconds subsequently.

Is this possible in anyway, instead of changing the `code` part of conf and using unique keys for map, I kinda don't want a separate key in logstash for each value as there can be many resources?

**Comment by fbaligand** [comment]
I understand what you expect. And unfortunately, this is not 'push_previous_map_as_event' behaviour.

In this option documentation (https://github.com/logstash-plugins/logstash-filter-aggregate#aggregate-plugin-options), there is an important warning : 

> WARNING: this option works fine only if tasks come one after the other. It means : all task1 events, then all task2 events, etc...

So, to have expected behavior, you must have the events in this order :
```
    {"resource_id": 1, "resource_type": "flight"}
    {"resource_id": 1, "resource_type": "flight"}
    {"resource_id": 2, "resource_type": "flight"}
    {"resource_id": 2, "resource_type": "flight"}
```

If it is not possible to change events order, there is an enhancement planned for your specific case, but not yet implemented : 
#33

**Comment by sameerr25** [comment]
Thanks @fbaligand.. When can we expect the 2.3.0 release with this enhancement?

**Comment by fbaligand** [comment]
I can't give you an exact date.
The PR is launched, I just started to review it.
So it could be soon.

**Comment by yaslamenace** [comment]
+1!
I am very interested i this feature as explained by sameerr25.
@fbaligand gand, Do you still have no idea about the date when the feature will be available ?

**Comment by fbaligand** [comment]
Hi guys, I can't promise you, but normally, I will release 2.3.0 version tomorrow with expected new feature !

**Comment by fbaligand** [comment]
@sameerr25 @yaslamenace 
release 2.3.0 is done with timeout event generation !

**Comment by fbaligand** [comment]
Close this issue, because fixed by #33

**Comment by yaslamenace** [comment]
Hi Fabien, Thank you for this new release.
However, i would like to know what they are the new functions that let me do what sameer and I are looking for. Indeed, there is nothing new in the explanation of the new release 2.3.0 (https://github.com/logstash-plugins/logstash-filter-aggregate)

Thanks !

**Comment by fbaligand** [comment]
Ok, so :
- new features are listed in changelog (which is referenced in home page) : 
https://github.com/logstash-plugins/logstash-filter-aggregate/blob/master/CHANGELOG.md
- example (# 3) explaining how to use these new features is also on home page : 
https://github.com/logstash-plugins/logstash-filter-aggregate#example-3--no-end-event

**Comment by yaslamenace** [comment]
Thanks Fabien, it works well !

**Comment by fbaligand** [comment]
Nice news !

**Comment by fbaligand** [comment]
For info, I just released version 2.3.1, with a new option :
*timeout_tags*, which let you define tags to add to generated event when a timeout occurs.
